single_kernel_mongo/managers/sharding.py

```diff
@@ -0,0 +1,689 @@
+        if not len(hosts):
+            logger.info(f"host info for shard {shard_name} not yet added, skipping")
+            return
+
+        self.charm.status_manager.to_maintenance(f"Adding shard {shard_name} to config-server")
+        config = self.state.mongos_config_for_user(
+            OperatorUser,
+            self.state.app_hosts,
+        )
+        with MongoConnection(config) as mongo:
+            try:
+                mongo.add_shard(shard_name, hosts)
+            except OperationFailure as e:
+                if e.code == 18:
+                    logger.error(
+                        f"{shard_name} shard does not have the same auth as the config server."
+                    )
+                    raise ShardAuthError(shard_name)
+            except PyMongoError as e:
+                logger.error(f"Failed to add {shard_name} to cluster")
+                raise e
+        self.charm.status_manager.to_active(None)
+
+    def remove_shards(self, relation: Relation):
+        """Removes a shard from the cluster."""
+        shard_name = relation.app.name
+
+        config = self.state.mongos_config_for_user(
+            OperatorUser,
+            self.state.app_hosts,
+        )
+        with MongoConnection(config) as mongo:
+            try:
+                self.charm.status_manager.to_maintenance(f"Draining shard {shard_name}")
+                logger.info("Attempting to removing shard: %s", shard_name)
+                mongo.pre_remove_shard_checks(shard_name)
+                mongo.remove_shard(shard_name)
+                mongo.move_primary_after_draining_shard(shard_name)
+            except NotReadyError:
+                logger.info("Unable to remove shard: %s another shard is draining", shard_name)
+                # to guarantee that shard that the currently draining shard, gets re-processed,
+                # do not raise immediately, instead at the end of removal processing.
+                raise ShardNotInClusterError
+            except ShardNotInClusterError:
+                logger.info(
+                    "Shard to remove is not in sharded cluster. It has been successfully removed."
+                )
+
+    def cluster_password_synced(self) -> bool:
+        """Returns True if the cluster password is synced."""
+        # base case: not config-server
+        if not self.state.is_role(MongoDBRoles.CONFIG_SERVER):
+            return True
+
+        config = self.state.mongos_config_for_user(
+            OperatorUser,
+            self.state.app_hosts,
+        )
+        try:
+            # check our ability to use connect to mongos
+            with MongoConnection(config) as mongos:
+                mongos.get_shard_members()
+            # check our ability to use connect to mongod
+            with MongoConnection(self.state.mongo_config) as mongod:
+                mongod.get_replset_status()
+        except OperationFailure as e:
+            if e.code in [13, 18]:  # Unauthorized, Auth Failed
```

Comment:
I've seen these numbers used elsewhere - it might be nice for these to be constants but it isn't necessary

Reply:
Mehdi asked the same in another PR and I'll bring an enum of the error codes that we handle in a future PR.